src/main/java/org/broadinstitute/hellbender/tools/copynumber/utils/optimization/PersistenceOptimizer.java

```diff
@@ -0,0 +1,208 @@
+
+        if (data.length == 1) {
+            return Collections.emptyList();
+        } else {
+            for (final int index : sortedIndices) {                             //watershed algorithm starts from lowest data point and iterates upward
+                if (index == 0) {                                               //first point has no left neighbor
+                    if (colors[index + 1] == NO_COLOR) {                        //local minimum, create new component
+                        createComponent(data, components, colors, index);
+                    } else {                                                    //extend component on right to include this point
+                        extendComponent(components, colors, colors[index + 1], index);
+                    }
+                    continue;
+                } else if (index == data.length - 1) {                          //last point has no right neighbor
+                    if (colors[index - 1] == NO_COLOR) {                        //local minimum, create new component
+                        createComponent(data, components, colors, index);
+                    } else {                                                    //extend component on left to include this point
+                        extendComponent(components, colors, colors[index - 1], index);
+                    }
+                    continue;
+                }
+
+                //otherwise, look both left and right
+                final int leftColor = colors[index - 1];
+                final int rightColor = colors[index + 1];
+                if (leftColor == NO_COLOR && rightColor == NO_COLOR) {          //local minimum, create new component
+                    createComponent(data, components, colors, index);
+                } else if (leftColor != NO_COLOR && rightColor == NO_COLOR) {   //extend component on left to include this point
+                    extendComponent(components, colors, leftColor, index);
+                } else if (leftColor == NO_COLOR && rightColor != NO_COLOR) {   //extend component on right to include this point
+                    extendComponent(components, colors, rightColor, index);
+                } else {                                                        //local maximum, merge components
+                    if (components.get(rightColor).minValue < components.get(leftColor).minValue) {
+                        extremaPairs.add(new ExtremaPair(data, components.get(leftColor).minIndex, index));
+                    } else {
+                        extremaPairs.add(new ExtremaPair(data, components.get(rightColor).minIndex, index));
+                    }
+                    mergeComponents(components, colors, leftColor, rightColor);
+                    colors[index] = colors[index - 1];                          //local maxmimum takes color of component to left by convention
```

Comment:
I'm a little confused here. Isn't the local maximum a part of a larger component at this point in the code, color of which is determined by which of the merged components has the lowest minimum? Also can't we just skip recoloring of the local maximum as we skip recoloring of non-boundary component values in the `mergeComponents` method?

Reply:
I don't think the maximum has been assigned a color yet, since we are traversing the points in ascending order, and both `createComponent` and `extendComponent` only assign colors to the point under consideration.

You are right in that assigning a color is not necessary just to find and order the local minima, but it's nice to have all points assigned a color at the end (and that's probably what would reasonably be expected from anyone looking to reuse this code in the future).

But perhaps it would be more consistent to also assign the color within `mergeComponents`; i.e., I'll change the signature of `mergeComponents` so that we call `mergeComponents(components, colors, leftColor, rightColor, index)` and will move the assignment of the color into the method accordingly.